### PR TITLE
Use optional realm attribute for authenticationrequest parameter max size/number validation configuration

### LINKED ISSUE #25006: Use optional realm attribute for authenticationrequest parameter max size/number validation configuration
### Description

1. It would be a security benefit if certain realms could have different i.e., more restricted values for the maximum size, and maximum number of additional parameters for an authentication request.
2. Also, the strategy to silently ignore additional request parameters that do not meet the requirements (number, length) is not suitable for everyone, as it does not enforce the fail-fast behavior.
3. Additional configuration support is also enabled by using a maximum size for all additional request parameter length combined.

1. The current values for `ADDITIONAL_REQ_PARAMS_MAX_MUMBER` is `5` and for `ADDITIONAL_REQ_PARAMS_MAX_SIZE` is `2000` (see https://github.com/keycloak/keycloak/blob/c0c0a80/services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java#L41).
2. The default behavior is still to silently ignore the additional request parameters. If the toggle is switched, an exception will be raised.
3. The default value is `Integer.MAX_VALUE`.

Configurable value enables customers to harden their keycloak instances realm specific by reducing the values according to their needs. All changes are backwards compatible and do NOT need any special configuration when updating keycloak the this version.

### Discussion

_No response_

### Motivation

This feature would enable customers to optionally configure the maximum number of additional parameters, the maximum size of each parameter, whether fail-fast strategy should be used and whether an overall parameter limit should be enforced, without having the need to change the keycloak-core.

### Details

The idea was to be definitely backwards compatible and don't force any configuration when upgrading to a new keycloak version.

### PR DESCRIPTION
**Pull request by mschallar**
This PR adds an optional configuration that enables realm specific authentication request parameter validation.

In detail, the maximum number and maximal length of authentication parameters (either provided as e.g., query, or via PAR) can be fine tuned via realm attribute. That enables customers to define realm specific validation i.e., hardening certain realms without having the need touch the keycloak-core base.
Additionally the fail-fast strategy can be toggled to give immediatelly feedback to the caller, whether the provided additional parameters meet the validation.

These changes doesn't change the default behaviour of parameter validation, hence it is fully backwards compatible.

Closes https://github.com/keycloak/keycloak/issues/25006

### CODE REVIEW

**Comment by sschu** [comment]
@mschallar If you are still interested in landing this, could you please address the comments I raised?

**Comment by mschallar** [comment]
@sschu You requested some tests, and I added some. Is there anything else I can provide?

**Comment by sschu** [comment]
@mschallar I made some inline comments on the code.

**Comment by mschallar** [comment]
Please point me to the comment on the code? I am unable to find it.

**Comment by sschu** [comment]
@mschallar My mistake, I submitted them as part of a review and forgot to release them. Sorry.

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java
Comment (mschallar) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java]: I clarified that condition, I hope it's now clearer. I had to use `>=` as, the parameter is added to `additionalReqParams` at the end of the method.

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java
Comment (mschallar) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java]: Yes, thanks

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java
Comment (mschallar) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java]: Yes, thanks

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java
Comment (mschallar) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java]: I added, and modified existing tests. I hope that by doing this everything is clear.

**Comment by mschallar** [comment]
I think that my last commit resolves all your comments.

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java
Comment (sschu) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java]: Actually, you are right. I didn't realise the new parameter is just added at the end. Its fine if you revert to your previous code, the new version is just longer and not really an improvement.

Comment (sschu) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java]: You are right. I did not realise this was already tested. You can re-add testing the skipping of wellknown params if you like (or maybe use a param with an empty value instead so this branch is also tested?). Maybe in a different position this time just to show it also works there.

Comment (mschallar) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java:203]: I added two additional well-known parameters to the text.

Comment (mschallar) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java:203]: Done, plus an comment for explanation

**Review by sschu — APPROVED** [review]
LGTM. Thanks! I'll try to get a review from the others. Maybe @keycloak/core-iam-maintainers ?

**Comment by stianst** [comment]
@keycloak/core-clients

**Comment by mschallar** [comment]
If there is anything else I can do to move this pull request forward, please let me know.

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java
Comment (mposolda) [inline services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java:206]: Is it possible to use `logger.warnv` instead? Usually `info` is supposed to be used for server-wide important information (EG. `Keycloak server successfully started in X seconds`) rather than error-like scenarios.

Similar likely applies to other `info` messages introduced in this PR.

**Review by mposolda — COMMENTED** [review]
@mschallar @sschu Thanks for the PR and review! It is nice contribution. Added one minor inline comment regarding logging.

Few other questions:

- Do we really need `additionalReqParamsMaxOverallSize` ? It seems to me that having 3 configuration properties is enough compromise between flexibility VS. too much complexity?

- Can we use server-wide configuration attributes instead of realm attributes? I afraid that if this is available by realm attributes, it may be needed to add the properties to the admin console as well (otherwise it is hidden and the only way to setup it would be to manually send admin REST request, which is not very convenient). But I don't know where to add those properties in the admin console... 
So IMO I would rather prefer those to be server-wide properties of some provider - maybe `oidc` protocol provider (See class `OIDCProviderConfig` class for the details). WDYT?

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java
Comment (mschallar) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/request/AuthzEndpointRequestParser.java:206]: As the client sent too many/too long request parameters, the KC client made a mistake, not the KC server. I think the KC server log should only contain an INFO that the client made a mistake, not a WARN. I would recommend to log to INFO rather to WARN.
A WARN log is always something the (keycloak) server owner should have a look at, as something may not be right on the server. That is certainly not the case here.

E.g.,
If the KC logfile is being monitored for WARN/ERROR logs, the monitor would alarm everytime a client does such a mistake. If the client contacts the keycloak owner about an error, the keycloak owner is still able to help/respond by checking the logs for the particilar keycloak client/timestamp.

**Comment by mschallar** [comment]
Thanks @mposolda
About your questions:

1. I incorporated the idea of a `additionalReqParamsMaxOverallSize` and `additionalReqParamsFailFast` from https://github.com/keycloak/keycloak/issues/16545 and thought "yea why not, if I do it let's do it properly". Hence, i designed the implementation in such a way that every configured parameter has a useful default- and backwards compatible value and does not need to be configured. 
2. True, we would have set those values via admin REST (ansible), or directly into the database via upgdate query. Unfortunately I am no expert in frontend, I cannot incorporate your idea into my PR if we want to keep it that way.
The alternative solution, a server-wide configuration sounds ok, but it would lose the charm of being realm specific (internal- vs. external realms may have different configurations). I will have a look into your example of `OIDCProviderConfig` soon, but I am pretty busy at the moment. Maybe we can still keep realm-specific configuration parameters that way as well.

**Comment by mschallar** [comment]
As we have not found any frontend developer that would bei interested in for the necessary frontend changes right now, I suggest we merge the code as it is, because:

- As all changes are backward compatible and do not require any manual configuration when updating, only advanced user will use those new configuration values. Those advanced users are also able to perform the correct (Rest-) Requests/Database changes to set those settings.
- I think if those options are more valuable to a broader audience in the future, someone will pick up your ideas @mposolda and update the GUI.

**Comment by mposolda** [comment]
@mschallar Thanks for the update. Can you please rebase the PR?

My personal preference is still to use rather server-wide configuration instead of realm attributes (since there is not GUI and hence attributes are somehow hidden). It may be possible to achieve this for example by:
- Re-introduce the class `OIDCProviderConfig`, which was removed in the https://github.com/keycloak/keycloak/commit/2d17024b14e9e55429556673ad19ed197e4ab1b8
- Make it possible on `OIDCLoginProtocol` class to have the method `public OIDCProviderConfig getConfig()`
- Obtain the configuration properties with something like 
```
OIDCProviderConfig config = keycloakSession.getProvider(OIDCLoginProtocol.class).getConfig();
this.additionalReqParamsMaxNumber = config.getAdditionalReqParamsMaxNumber();  
```
instead of realm attributes.

Also I suggest to change the log level either to `debug` or to `warn`, but rather not to `info` as `info` is really dedicated for the informational messages related to whole server, but not to the user-level errors (like user providing incorrect parameters).

**Comment by patrick-primesign** [comment]
@mposolda I rebased @mschallar PR.

I will also look into your proposed changes with re-adding OIDCProviderConfig and regarding logging.

**Comment by mposolda** [comment]
@patrick-primesign Thanks!

**Comment by patrick-primesign** [comment]
@mposolda I rebased again and made the changes. 

I added these settings to the http options, altough I'm not sure if it is the right place for them.

The settings could be set  eg. like this: 
`kc.sh --spi-login-protocol-openid-connect-add-req-params-max-size=2000`
Or via environment:
`KC_SPI_LOGIN_PROTOCOL_OPENID_CONNECT_ADD_REQ_PARAMS_MAX_SIZE=2000`

Thanks for your time.

**Review by mposolda — COMMENTED** [review]
Thanks for the update! The PR looks much better now IMO.I've added few comments inline. The most important ones:
- AFAIK the changes in `HttpOptions` and `HttpPropertyMappers` are not needed and can be reverted from this PR
- Please make sure to setup your IDE to not use tabs, but instead use 4 spaces indentation similarly like all other Keycloak sources
- Few more minor comments regarding the test optimization and small javadoc update

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/authz/AuthzEndpointRequestParserTest.java
Comment (patrick-primesign) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/authz/AuthzEndpointRequestParserTest.java:23]: I left it as a empty method, in the AbstractKeycloakTest I would have to implement addTestRealms which would have ended in me copying the AbstractTestRealmKeycloakTest.

**Comment by patrick-primesign** [comment]
@mposolda Thanks again for looking over the code. Hopefully now everything is ok and it still works as intended.

**Review by mposolda — DISMISSED** [review]
@patrick-primesign Thanks for the updates!

I hope this can be merged once tests are OK

**Comment by mposolda** [comment]
@patrick-primesign Could you please rebase this PR? As there are some conflicts. I hope this is the last step to make this PR ready for merge.

**Comment by patrick-primesign** [comment]
@mposolda Great thanks!
I just rebased, so there should be no problem.

**Review by mposolda — APPROVED** [review]

**Pull request merged**
